Firestore/core/src/firebase/firestore/core/filter.h

```diff
@@ -67,8 +77,30 @@ class Filter {
 
   /** A unique ID identifying the filter; used when serializing queries. */
   virtual std::string CanonicalId() const = 0;
+
+  /** A debug description of the Filter. */
+  virtual std::string ToString() const = 0;
+
+  virtual size_t Hash() const = 0;
+
+  virtual bool IsInequality() const {
+    return false;
+  }
+
+  friend bool operator==(const Filter& lhs, const Filter& rhs) {
+    return lhs.Equals(rhs);
+  }
+
+ protected:
+  virtual bool Equals(const Filter& other) const = 0;
```

Comment:
Optional: I think this function can be `private` (I don't think it's ever called by the derived classes).

Reply:
I think it's a fairly well-known idiom -- Sutter is a big proponent.